README.md
# php-apm

CHANGELOG.md
# Package php-apm Change Log

src/Test.php
<?php
declare(strict_types=1);
namespace Middleware\PhpApmTest2;

require 'vendor/autoload.php';

class Test {
    public static function printString($str): void {
        echo $str;
    }

    public static function printServerDetails($serverJson, $key = ''): void {
        if ($key != '' && isset($serverJson[$key]) && !empty($serverJson[$key])) {
            echo '<pre>';
            print_r($serverJson[$key]);
            echo '</pre>';
        } else {
            echo '<pre>';
            print_r($serverJson);
            echo '</pre>';
        }
    }
}

//Test::printString('Hello..called. from DemoClass');


Example.php
<?php
declare(strict_types=1);

require 'vendor/autoload.php';
use Middleware\PhpApmTest\PhpApmCollector;


class Test {
    public static function printString($str): void {
        echo $str . PHP_EOL;
    }
}

//Test::printString('Hello..called. from Outside.');

class DemoClass {
    public static function runCode(): void {
        $mwCollector = new PhpApmCollector('test-project-51', 'test-service-51');
        $mwCollector->tracingCall(get_called_class(), __FUNCTION__, __FILE__, __LINE__);
        // $mwCollector->preTracingCall(get_called_class(), __FUNCTION__, __FILE__, __LINE__);

        Test::printString('Hello..called. from DemoClass' . PHP_EOL);
        // $mwCollector->postTracingCall();

    }
}

class TestClass {
    public static function printString(): void {
        $mwCollector = new PhpApmCollector('test-project-52', 'test-service-52');
        $mwCollector->tracingCall(get_called_class(), __FUNCTION__, __FILE__, __LINE__);

        Test::printString('Hello..called. from TestClass' . PHP_EOL);

    }
}

DemoClass::runCode();
TestClass::printString();

src/LoopTest.php
<?php
//https://github.com/SigNoz/sample-php-app/blob/a51a65390bc628a2b61dcda86cbf3c9bc3c6ce62/src/2-send-trace-to-collector.php

declare(strict_types=1);
require __DIR__ . '/../vendor/autoload.php';

use GuzzleHttp\Client;
use GuzzleHttp\Psr7\HttpFactory;
use OpenTelemetry\Contrib\Otlp\OtlpHttpTransportFactory;
use OpenTelemetry\Contrib\Otlp\SpanExporter;
use OpenTelemetry\SDK\Trace\SpanProcessor\SimpleSpanProcessor;

use OpenTelemetry\SDK\Trace\Tracer;
use OpenTelemetry\SDK\Trace\TracerProvider;
use OpenTelemetry\API\Trace\SpanKind;
use OpenTelemetry\SDK\Trace\TracerProviderFactory;
use OpenTelemetry\SDK\Common\Attribute\Attributes;

//putenv('OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:9321/v1/traces');
//putenv('OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf');
putenv('OTEL_PHP_AUTOLOAD_ENABLED=true');
putenv('OTEL_SERVICE_NAME=mw-php-app');

echo 'Starting CollectorSpanExporter' . PHP_EOL;

$transport = (new OtlpHttpTransportFactory())->create('http://localhost:9321/v1/traces', 'application/x-protobuf');
$exporter = new SpanExporter($transport);

$tracerProvider = new TracerProvider(
    new SimpleSpanProcessor(
        $exporter
    )
);

$tracer = $tracerProvider->getTracer('io.opentelemetry.contrib.php');

$root = $span = $tracer->spanBuilder('root')->startSpan();
$rootScope = $span->activate();

for ($i = 0; $i < 3; $i++) {
    // start a span, register some events
    $span = $tracer->spanBuilder('loop-' . $i)->startSpan();

    $span->setAttribute('remote_ip', '1.2.3.4')
        ->setAttribute('country', 'USA');

    $span->addEvent('found_login' . $i, [
        'id' => $i,
        'username' => 'otuser' . $i,
    ]);
    $span->addEvent('generated_session', [
        'id' => md5((string) microtime(true)),
    ]);

    $span->end();
}

$root->end();
$rootScope->detach();
echo PHP_EOL . 'CollectorSpanExporter complete!  ';

echo PHP_EOL;

src/OtelHook.php
<?php
declare(strict_types=1);
namespace Middleware\PhpApmTest;

//"open-telemetry/sdk": "^1.0.0beta3",
//"open-telemetry/api": "1.0.0beta4",

//https://opentelemetry.io/docs/instrumentation/php/automatic/

//https://github.com/open-telemetry/opentelemetry-php-instrumentation

//https://morioh.com/p/d1f7b1aed614

require __DIR__ . '/../vendor/autoload.php';

use GuzzleHttp\Client;
use GuzzleHttp\Psr7\HttpFactory;
use OpenTelemetry\Contrib\Otlp\OtlpHttpTransportFactory;
use OpenTelemetry\Contrib\Otlp\SpanExporter;
use OpenTelemetry\SDK\Trace\SpanProcessor\SimpleSpanProcessor;

use OpenTelemetry\SDK\Trace\TracerProvider;
use OpenTelemetry\SDK\Common\Instrumentation As Instrumentation;

//putenv('OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:9321/v1/traces');
//putenv('OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf');
putenv('OTEL_PHP_AUTOLOAD_ENABLED=true');
putenv('OTEL_SERVICE_NAME=mw-php-app');

echo 'Starting CollectorSpanExporter' . PHP_EOL;

class DemoClass {
    public static function run() {
        echo 'Hello..called.';
        $transport = (new OtlpHttpTransportFactory())->create('http://localhost:9321/v1/traces', 'application/x-protobuf');
        $exporter = new SpanExporter($transport);

        $tracerProvider = new TracerProvider(
            new SimpleSpanProcessor(
                $exporter
            )
        );

        //$tracer = $tracerProvider->getTracer('io.opentelemetry.contrib.php');
        $tracer = new Tracer($tracerProvider);

        Instrumentation\hook(
            DemoClass::class,
            'run',
            static function (DemoClass $demo, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($tracer) {
                /*static $instrumentation;
                $instrumentation ??= new CachedInstrumentation('example');
                $instrumentation->tracer()->spanBuilder($class)
                    ->startSpan()
                    ->activate();*/

                $span = $tracer->spanBuilder($class)
                    ->startSpan()
                    ->activate();
                Context::storage()->attach($span->storeInContext(Context::getCurrent()));
            },
            static function (DemoClass $demo, array $params, $returnValue, ?Throwable $exception) use ($tracer) {
                $scope = Context::storage()->scope();
                $scope->detach();
                $span = Span::fromContext($scope->context());
                if ($exception) {
                    $span->recordException($exception);
                    $span->setStatus(StatusCode::STATUS_ERROR);
                }
                $span->end();
            }
        );

        $demo = new DemoClass();
        $demo->run();
    }
}


PhpApmCollector.php
<?php
declare(strict_types=1);
namespace Middleware\PhpApmTest;

require 'vendor/autoload.php';

use OpenTelemetry\Context\Context;
use OpenTelemetry\API\Trace\StatusCode;
use OpenTelemetry\Contrib\Otlp\OtlpHttpTransportFactory;
use OpenTelemetry\Contrib\Otlp\SpanExporter;
use OpenTelemetry\SDK\Common\Attribute\Attributes;
use OpenTelemetry\SDK\Common\Configuration\Variables;
use OpenTelemetry\SDK\Resource\ResourceInfo;
use OpenTelemetry\Sdk\Trace\Span;
use OpenTelemetry\SDK\Trace\SpanProcessor\SimpleSpanProcessor;
use OpenTelemetry\SDK\Trace\TracerProvider;
use OpenTelemetry\API\Trace\TracerInterface;
use OpenTelemetry\SemConv\ResourceAttributes;

final class PhpApmCollector {

    private int $exportPort = 9321;
    private string $projectName;
    private string $serviceName;
    private TracerInterface $tracer;

    public function __construct(string $projectName = null, string $serviceName = null) {

        if (empty($projectName)) {
            $projectName = 'Project-'. getmypid();
        }

        if (empty($serviceName)) {
            $serviceName = 'Service-'. getmypid();
        }

        $this->projectName = $projectName;
        $this->serviceName = $serviceName;

        $transport = (new OtlpHttpTransportFactory())->create(
            'http://localhost:' . $this->exportPort . '/v1/traces',
            'application/x-protobuf');

        $exporter = new SpanExporter($transport);

        $tracerProvider = new TracerProvider(
            new SimpleSpanProcessor($exporter),
            null,
            ResourceInfo::create(Attributes::create([
                'project.name' => $projectName,
                ResourceAttributes::SERVICE_NAME => $serviceName,
                Variables::OTEL_PHP_AUTOLOAD_ENABLED => true,
            ]))
        );

        $this->tracer = $tracerProvider->getTracer('io.opentelemetry.contrib.php', '0.1.0');
    }

    public function preTracingCall(
        ?string $className,
        string $functionName,
        ?string $fileName,
        ?iterable $attributes = null): void {
        $span = $this->tracer->spanBuilder(sprintf('%s::%s', $className, $functionName))
            ->setAttribute('service.name', $this->serviceName)
            ->setAttribute('project.name', $this->projectName)
            ->setAttribute('code.function', $functionName)
            ->setAttribute('code.namespace', $className)
            ->setAttribute('code.filepath', $fileName);

        if (!empty($attributes)) {
            foreach ($attributes as $key => $value) {
                $span->setAttribute($key, $value);
            }
        }
        $span = $span->startSpan();

        // $span = $tracer
        //     ->spanBuilder(sprintf('%s %s', $request->getMethod(), $request->getUri()))
        //     ->setSpanKind(SpanKind::KIND_CLIENT)
        //     ->setAttribute('http.method', $request->getMethod())
        //     ->setAttribute('http.url', $request->getUri())
        //     ->startSpan();
        // $span = $tracer
        //     ->spanBuilder('get-user')
        //     ->setAttribute('db.system', 'mysql')
        //     ->setAttribute('db.name', 'users')
        //     ->setAttribute('db.user', 'some_user')
        //     ->setAttribute('db.statement', 'select * from users where username = :1')
        //     ->startSpan();
        Context::storage()->attach($span->storeInContext(Context::getCurrent()));
    }

    public function postTracingCall(): void {
        if (!$scope = Context::storage()->scope()) {
            return;
        }

        $scope->detach();
        $span = Span::fromContext($scope->context());
        $span->setStatus(StatusCode::STATUS_OK);
        $span->end();
    }

    public function tracingCall(
        ?string $className,
        string $functionName,
        ?string $fileName,
        ?iterable $attributes = null): void {
        $this->preTracingCall($className, $functionName, $fileName, $attributes);
        $this->postTracingCall();
    }
}

src/GettingStartedConsole.php
<?php

declare(strict_types=1);
require __DIR__ . '/../vendor/autoload.php';

use OpenTelemetry\SDK\Trace\SpanExporter\ConsoleSpanExporterFactory;
use OpenTelemetry\SDK\Trace\SpanProcessor\SimpleSpanProcessor;
use OpenTelemetry\SDK\Trace\TracerProvider;

echo 'Starting ConsoleSpanExporter' . PHP_EOL;

$tracerProvider = new TracerProvider(
    new SimpleSpanProcessor(
        (new ConsoleSpanExporterFactory())->create()
    )
);

$tracer = $tracerProvider->getTracer('io.opentelemetry.contrib.php');

$rootSpan = $tracer->spanBuilder('root')->startSpan();
$rootScope = $rootSpan->activate();

try {
    $span1 = $tracer->spanBuilder('foo')->startSpan();
    $scope = $span1->activate();
    try {
        $span2 = $tracer->spanBuilder('bar')->startSpan();
        echo 'OpenTelemetry welcomes PHP' . PHP_EOL;
    } finally {
        $span2->end();
    }
} finally {
    $span1->end();
    $scope->detach();
}
$rootSpan->end();
$rootScope->detach();

src/GettingStartedCollector.php
<?php

declare(strict_types=1);
require __DIR__ . '/../vendor/autoload.php';

use GuzzleHttp\Client;
use GuzzleHttp\Psr7\HttpFactory;
use OpenTelemetry\Contrib\Otlp\OtlpHttpTransportFactory;
use OpenTelemetry\Contrib\Otlp\SpanExporter;
use OpenTelemetry\SDK\Trace\SpanExporter\ConsoleSpanExporter;
use OpenTelemetry\SDK\Trace\SpanProcessor\SimpleSpanProcessor;
use OpenTelemetry\SDK\Trace\TracerProvider;

echo 'Starting OTLP/http Exporter' . PHP_EOL;

//$transport = (new OtlpHttpTransportFactory())->create('http://collector:4318/v1/traces', 'application/x-protobuf');
$transport = (new OtlpHttpTransportFactory())->create('http://localhost:9321/v1/traces', 'application/x-protobuf');
$exporter = new SpanExporter($transport);

$tracerProvider = new TracerProvider(
    new SimpleSpanProcessor(
        $exporter
    )
);

$tracer = $tracerProvider->getTracer('io.opentelemetry.contrib.php');

$rootSpan = $tracer->spanBuilder('root')->startSpan();
$rootScope = $rootSpan->activate();

try {
    $span1 = $tracer->spanBuilder('foo')->startSpan();
    $scope = $span1->activate();
    try {
        $span2 = $tracer->spanBuilder('bar')->startSpan();
        echo 'OpenTelemetry welcomes PHP' . PHP_EOL;
    } finally {
        $span2->end();
    }
} finally {
    $span1->end();
    $scope->detach();
}
$rootSpan->end();
$rootScope->detach();

TestOtel.php
<?php
declare(strict_types=1);

require 'vendor/autoload.php';
use Middleware\PhpApmTest\PhpApmCollector;


class Test {
    public static function printString($str): void {
        echo $str . PHP_EOL;
    }
}

//Test::printString('Hello..called. from Outside.');

class DemoClass {
    public static function runCode(): void {
        $mwCollector = new PhpApmCollector('test-project-51', 'test-service-51');
        $mwCollector->tracingCall(get_called_class(), __FUNCTION__, __FILE__, __LINE__);
        // $mwCollector->preTracingCall('apbb123', get_called_class(), __FUNCTION__, __FILE__, __LINE__);

        Test::printString('Hello..called. from DemoClass' . PHP_EOL);
        // $mwCollector->postTracingCall();

    }
}

class TestClass {
    public static function printString(): void {
        $mwCollector = new PhpApmCollector('test-project-52', 'test-service-52');
        $mwCollector->tracingCall(get_called_class(), __FUNCTION__, __FILE__, __LINE__);

        Test::printString('Hello..called. from TestClass' . PHP_EOL);

    }
}

DemoClass::runCode();
TestClass::printString();